watchtower/wtclient/interface.go

```diff
@@ -95,21 +96,18 @@ type DB interface {
 	AckUpdate(id *wtdb.SessionID, seqNum, lastApplied uint16) error
 }
 
-// Dial connects to an addr using the specified net and returns the connection
-// object.
-type Dial func(net, addr string) (net.Conn, error)
-
 // AuthDialer connects to a remote node using an authenticated transport, such as
 // brontide. The dialer argument is used to specify a resolver, which allows
 // this method to be used over Tor or clear net connections.
-type AuthDialer func(localKey keychain.SingleKeyECDH, netAddr *lnwire.NetAddress,
-	dialer func(string, string) (net.Conn, error)) (wtserver.Peer, error)
+type AuthDialer func(
+	localKey keychain.SingleKeyECDH, netAddr *lnwire.NetAddress,
+	dialer tor.DialFunc) (wtserver.Peer, error)
 
 // AuthDial is the watchtower client's default method of dialing.
 func AuthDial(localKey keychain.SingleKeyECDH, netAddr *lnwire.NetAddress,
-	dialer func(string, string) (net.Conn, error)) (wtserver.Peer, error) {
+	dialer tor.DialFunc) (wtserver.Peer, error) {
 
-	return brontide.Dial(localKey, netAddr, dialer)
+	return brontide.Dial(localKey, netAddr, tor.DefaultConnTimeout, dialer)
```

Comment:
Sometimes we're using the default, and sometimes we're using the value set in config. Is that just so that we don't need to make more interface changes? Seems strange to set a connection timeout in config, but then only have it sometimes apply.

Reply:
I was actually not sure about this change, some suggestions would really be appreciated.
With the timeout being built into the `Dial` function, all the related interfaces now have to be changed to take the new parameter. Currently two services are using the default value, `bootstrapper` and `watchtower`, it was done this way to have least code change, but yeah it is indeed weird... we can of course pass the config value to the services, then,
 - the `DNSSeedBootstrapper` struct would need a new field to store the timeout value.
 - the `watchtower` will change the function `AuthDial` and the type `AuthDialer`, which will cause more changes to the underlying functions.

It was doable, while I feel the original design was better and easier, that we only touch the `tor.net`. And to solve the race condition, we make a second dial function, `DialTimeout`, on both `brontide` and `tor.Net` interface. For any requests want a customized timeout value, use the `DialTimeout`. What do you think?

Reply:
> the DNSSeedBootstrapper struct would need a new field to store the timeout value.

I think this is ok, minimal change that lets us thread it through cleanly. 

> authDial and the type AuthDialer, which will cause more changes to the underlying functions

It would be nice to leave these untouched as you say, so we could leave watchtower's `AuthDial` signature unchanged, and then just wrap the call to `brontide.Dial` where we set it in the [server](https://github.com/lightningnetwork/lnd/blob/master/server.go#L1210). Since the existing function is such a thin wrapper already.

Reply:
Sure, the wrapper helps. Fixed according to your suggestions, thanks!